system-test/trace-express.js

```diff
@@ -0,0 +1,129 @@
+    'GOOGLE_APPLICATION_CREDENTIALS');
+}
+
+// trace-agent must be loaded before everything else.
+require('../').start({ flushDelaySeconds: 1 });
+
+const assert = require('assert');
+const googleAuth = require('google-auto-auth');
+const got = require('got');
+const queryString = require('querystring');
+const uuid = require('uuid');
+
+const WRITE_CONSISTENCY_DELAY_MS = 20 * 1000;
+const projectId = process.env.GCLOUD_PROJECT;
+
+function makeAuthorizedGot(auth) {
+  function authorize(options) {
+    return new Promise((resolve, reject) => {
+      auth.authorizeRequest(options, (err, authorizedOptions) => {
+        if (err) {
+          return reject(err);
+        }
+        return resolve(authorizedOptions);
+      });
+    });
+  }
+
+  return (url, options) => {
+    return authorize(options)
+      .then((authorizedOptions) => {
+        return got(url, authorizedOptions);
+      });
+  };
+}
+
+function listTraces(testPath) {
+  const BASE_URI = `https://cloudtrace.googleapis.com/v1/projects/${projectId}`;
+  const auth = googleAuth({
+    scopes: ['https://www.googleapis.com/auth/trace.readonly']
+  });
+  const agot = makeAuthorizedGot(auth);
+
+  const query = queryString.stringify({
+    filter: `span:${testPath}`
+  });
+  const uri = `${BASE_URI}/traces?${query}`;
+
+  return agot(uri, { json: true }).then((response) => {
+    const body = response.body;
+    const promises = body.traces.map((trace) => {
+      const uri = `${BASE_URI}/traces/${trace.traceId}`;
+      return agot(uri, {
+        json: true
+      });
+    });
+
+    return Promise.all(promises).then((responses) => {
+      return responses.map(response => response.body);
+    });
+  });
+}
+
+describe('express + datastore', () => {
+  it('should be able to trace datastore (grpc) calls', (done) => {
+    // Build a unique path so that we get unique trace span names.
+    const testPath = `/test-${uuid.v4()}`;
+
+    const express = require('express');
+    const datastore = require('@google-cloud/datastore')();
+
+    const app = express();
+    app.get(testPath, (req, res) => {
+      // Make a query to a non-existent datastore entity. This will get traced
+      // regardless.
+      const query = datastore.createQuery('Task').order('created');
+      datastore.runQuery(query)
+        .then((results) => {
+          console.log('datastore results:', results);
+          res.status(200).send('hello\n');
+        });
+    });
+
+    var server = app.listen(8080, () => {
+      console.log('server started');
+      got.stream(`http://localhost:8080${testPath}`).pipe(process.stdout);
+      setTimeout(() => {
+        listTraces(testPath).then(verifyTraces);
+      }, WRITE_CONSISTENCY_DELAY_MS);
+    });
+
+    function verifyTraces(traces) {
+      assert.equal(traces.length, 1, 'there should be exactly one trace');
+
+      const trace = traces[0];
+      console.log(trace);
+      assert.equal(trace.spans.length, 2, 'should be 2 spans: parent, child');
```

Comment:
Should be fully deterministic. Not sure why you wouldn't but this was more just a heads up in case you observed more than 2 spans.